### ISSUE TITLE
Installation issue: `openblas` does not have a correct `install_name` on macos

### ISSUE THREAD

**Issue by Chrismarsh** [LLNL/spack#45464]
### Error message

`openblas` has an `install_name` of `libopenblas.0.dylib` which means that some consuming packages struggle to consume libraries linked against openblas.

For example, `py-numpy` links against `openblas`, but when `py-numpy` is used by `py-netcdf4`, `py-netcdf4` fails to build as numpy does not prefix `@rpath/`. 

```
$ otool -L /Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so
/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so:
	libopenblas.0.dylib (compatibility version 0.0.0, current version 0.3.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.100.2)
``` 

```
[...]
          cmd LC_RPATH
      cmdsize 160
         path /Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-oecxneqr7e2rdzlmv36r6wq4dn5aac2j/lib (offset 12)
Load command 26
```

```
$ spack install py-netcdf4
[...]

  Traceback (most recent call last):
    File "/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/__init__.py", line 24, in <module>
      from . import multiarray
    File "/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/multiarray.py", line 10, in <module>
      from . import overrides
    File "/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/overrides.py", line 8, in <module>
      from numpy.core._multiarray_umath import (
  ImportError: dlopen(/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so, 0x0002): Library not loaded: libopenblas.0.dylib
    Referenced from: <96301193-7C5B-311E-9A15-E84A444ECE1C> /Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/py-numpy-1.26.4-47jedfkjvehucaeawa47zz5xm7vkogjd/lib/python3.11/site-packages/numpy/core/_multiarray_umath.cpython-311-darwin.so
    Reason: tried: 'libopenblas.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibopenblas.0.dylib' (no such file), 'libopenblas.0.dylib' (no such file), '/Users/cmarsh/Documents/science/code/spack/stage/spack-stage-py-netcdf4-1.6.5-r52uf346tmcn7fmhko3phfo5vnztmt67/spack-src/libopenblas.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/stage/spack-stage-py-netcdf4-1.6.5-r52uf346tmcn7fmhko3phfo5vnztmt67/spack-src/libopenblas.0.dylib' (no such file), '/Users/cmarsh/Documents/science/code/spack/stage/spack-stage-py-netcdf4-1.6.5-r52uf346tmcn7fmhko3phfo5vnztmt67/spack-src/libopenblas.0.dylib' (no such file)
```

/cc @adamjstewart 
Adam, this is the root of the build errors I reported in Slack yesterday.

I think the simplest solution is to ensure openblas has a sane install_name.

### Information on your system

```
spack debug report
* **Spack:** 0.23.0.dev0 (4240748ceac4eafd2e1b34dd98a93db41cbf5796)
* **Python:** 3.12.3
* **Platform:** darwin-sonoma-m1
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by adamjstewart** [comment]
Your fix in https://github.com/Chrismarsh/spack/commit/185dc22b71a63e823f618225da24062fd28b5c4d looks right, but should probably be applied to makefile, not cmake. I don't normally see this issue with cmake. Want to open a PR?

**Comment by Chrismarsh** [comment]
Ya I was surprised too.

```
$ spack spec /oo2xo5qoyvsop3zyogpq3ngllgv7il2r
Input spec
--------------------------------
 -   /oo2xo5qoyvsop3zyogpq3ngllgv7il2r

Concretized
--------------------------------
[+]  openblas@0.3.27%apple-clang@15.0.0~bignuma~consistent_fpcsr+dynamic_dispatch+fortran~ilp64~ipo+locking+pic+shared build_system=cmake build_type=Release generator=make symbol_suffix=none threads=none arch=darwin-sonoma-m1
[+]      ^cmake@3.30.1%apple-clang@15.0.0~doc+ncurses+ownlibs build_system=generic build_type=Release patches=dbc3892 arch=darwin-sonoma-m1

$ otool -D libopenblas.0.3.dylib
libopenblas.0.3.dylib:
libopenblas.0.dylib
```

I can't get the makefile variant to build -- I think it's pikcing up homebrew's gfortran so I cannot test what its output looks like.
```
==> Error: ProcessError: Command exited with status 2:
    'make' '-j10' '-s' 'CC=/Users/cmarsh/Documents/science/code/spack/lib/spack/env/clang/clang' 'FC=/Users/cmarsh/Documents/science/code/spack/lib/spack/env/clang/gfortran' 'MAKE_NB_JOBS=0' 'ARCH=arm64' 'DYNAMIC_ARCH=1' 'DYNAMIC_OLDER=1' 'TARGET=GENERIC' 'DYNAMIC_ARCH=1' 'USE_LOCKING=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'RANLIB=ranlib' 'all'

2 errors found in build log:
     17340    ld: warning: could not create compact unwind for _dlatm1_: does not use standard frame
     17341    ld: warning: could not create compact unwind for _dlatm7_: does not use standard frame
     17342    ld: warning: could not create compact unwind for _dlarnd_: does not use standard frame
     17343    dyld[31106]: Library not loaded: @rpath/libgfortran.5.dylib
     17344      Referenced from: <4DE0C938-0F1A-33E0-89CA-8C125666D516> /Users/cmarsh/Documents/science/code/spack/stage/spack-stage-openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/spack-src/utest/openblas_utest
     17345      Reason: tried: '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no
              such file), '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no s
              uch file), '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib/libgfortran.5.dylib' (no such f
              ile), '/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-asrysdsjcx6zr7hxrsf5igpep2xe7ty6/lib64/libgfortran.5.dylib' (no such fi
              le)
  >> 17346    make[1]: *** [Makefile:81: run_test] Abort trap: 6
  >> 17347    make: *** [Makefile:168: tests] Error 2

```

**Comment by Chrismarsh** [comment]
It was not homebrew's fault; rather the issue about gfortran rpaths detailed in #45479. 

I can now build `openblas build_system=makefile` and I can confirm this problem _only_ affects the cmake build:

```
$ spack spec /6tmtmod6xmgqxwza3osgiopkuowynufr
Input spec
--------------------------------
 -   /6tmtmod6xmgqxwza3osgiopkuowynufr

Concretized
--------------------------------
[+]  openblas@0.3.27%apple-clang@15.0.0~bignuma~consistent_fpcsr+dynamic_dispatch+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=none arch=darwin-sonoma-m1
[...]

$ otool -D libopenblas.dylib
libopenblas.dylib:
/Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-6tmtmod6xmgqxwza3osgiopkuowynufr/lib/libopenblas.0.dylib
```

**Comment by adamjstewart** [comment]
Still so surprised CMake has this issue at all. Maybe open an issue on the OpenBLAS GitHub to get to the bottom of this? Or ping one of our CMake maintainers. At this point I'm just really curious. I know you've already done an insane amount of debugging though.

**Comment by Chrismarsh** [comment]
I've opened an issue upstream here:
https://github.com/OpenMathLib/OpenBLAS/issues/4823
it looks like the makefile explicitly sets the `install_name` but the cmake build doesn't. Openblas cmake doesn't set `MACOSX_RPATH=ON`, so I think this `install_name` is as expected.

**Comment by adamjstewart** [comment]
But `CMakePackage` sets `CMAKE_INSTALL_RPATH_USE_LINK_PATH`, shouldn't that work?

**Comment by Chrismarsh** [comment]
No, I don't think so. 

`CMAKE_INSTALL_RPATH_USE_LINK_PATH` sets the binary [to have a set of `rpath`s](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_RPATH_USE_LINK_PATH.html) that correspond to the paths used during link. And indeed, the openblas rpath is added to libraries that use openblas:
```
[...]
          cmd LC_RPATH
      cmdsize 160
         path /Users/cmarsh/Documents/science/code/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/openblas-0.3.27-oecxneqr7e2rdzlmv36r6wq4dn5aac2j/lib (offset 12)
Load command 26
```

But: [as I understand it](https://developer.apple.com/documentation/xcode/build-settings-reference#Dynamic-Library-Install-Name), the `install_name` of the dylib is what is used during link to control how the consuming binary finds the library. The `install_name` can be: a) a relative path, b) an absolute path, c) or a special (e.g., `@rpath/mylib.dylib`). Because the openblas `install_name` is just a name, binaries linked against it it cannot find it. The `install_name` of the openblas dylib needs to be `@rpath/libopenblas.0.dylib` for it to work with `CMAKE_INSTALL_RPATH_USE_LINK_PATH`.

This is a nice description, which is what I have used to understand this over the years
https://web.archive.org/web/20200814225809/http://log.zyxar.com/blog/2012/03/10/install-name-on-os-x/

**Comment by Chrismarsh** [comment]
@adamjstewart this has been resolved upstream for `@0.3.28:`. I have added a commented and filed the PR https://github.com/spack/spack/pull/45606
